R/creds_helpers.R

```diff
@@ -0,0 +1,83 @@
+#' Helpers for supplying SMTP credentials
+#'
+#' These helper functions, the credential helpers, are used to supply SMTP
+#' configuration and authorization information for the [smtp_send()] function.
+#' These helpers ([creds_file()], [creds_key()], and [creds()]) are to be used
+#' expressly with the `credentials` argument of [smtp_send()]. The
+#' [creds_file()] credential helper is used to point at a credentials file
+#' stored on disk. We can create that file using the [create_smtp_creds_file()]
+#' function. The [creds_key()] credential helper is truly somthing special. It
+#' actually helps one obtain credentials that are stored in the system-wide
+#' key-value store. The equally magical [create_smtp_creds_key()] function sets
+#' that key. Using [creds()] allows for manual specification of SMTP
+#' configuration and authentication within that helper function.
+#'
+#' @param file When using the [creds_file()] credential helper, we need to
+#'   specify the location of the credential file, and, this is where that is
+#'   done. The credential file was ideally generated by the
+#'   [create_smtp_creds_file()] function.
+#' @param key_name When using the [creds_key()] credential helper, the name of
+#'   the key (in the system key-value store) needs to be given here. This was
+#'   either explicitly provided when using the [create_smtp_creds_key()]
+#'   function (with its own `key_name` argument), or, it may have been
+#'   automatically generated based on the SMTP host name.
+#' @inheritParams create_smtp_creds_file
+#' @name credential_helpers
+#' @return a list object of class `location_cells`.
+NULL
+
+#' Credentials helper for pointing to a SMTP credentials file
+#'
+#' @rdname credential_helpers
+#' @export
+creds_file <- function(file) {
+
+  structure(
+    class = c("creds_file", "blastula_creds"),
+    list(
+      file = file
+    )
+  )
+}
+
+#' Credentials helper for pointing to a SMTP credentials key
+#'
+#' @rdname credential_helpers
+#' @export
+creds_key <- function(key_name) {
+
+  structure(
+    class = c("creds_key", "blastula_creds"),
+    list(
+      key_name = key_name
+    )
+  )
+}
+
+#' Credentials helper for manual entry of SMTP config and auth values
+#'
+#' @rdname credential_helpers
+#' @export
+creds <- function(provider = NULL,
```

Comment:
1. Introduce a new `creds_internal` or something, that is not exported, and takes a `password` parameter. It won't call `create_credentials_list` because that prompts for password.
2. `creds` should pass through all arguments to `creds_internal`, and pass `password = getpass()` or whatever. Or it can call `create_credentials_list`.
3. `creds_key` and `creds_file` should call `creds_internal`.

The point being that the returned object from `creds_key`, `creds_file`, and `creds` should look identical.

That object should have an overridden `print` method that redacts the password. (Maybe `str`/`format` as well? Ask Barret or Winston.)

Reply:
And `creds_anonymous` that doesn't have a `user` and doesn't prompt for password.

Reply:
Now done. Thanks for the detailed instructions on implementation.